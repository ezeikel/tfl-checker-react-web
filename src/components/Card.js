import React, { useState } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Slider from "react-slick";
import "slick-carousel/slick/slick.css";
import "slick-carousel/slick/slick-theme.css";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const Wrapper = styled.section`
  display: flex;
  flex-direction: column;
  height: 300px;
  box-shadow: var(--box-shadow);
  border-radius: var(--border-radius);
`;

const Name = styled.div`
  flex: 0 0 50%;
  background-color: ${props => `var(--color-${props.id})`};
  color: var(--color-white);
  display: flex;
  align-content: center;
  justify-content: center;
  border-top-left-radius: var(--border-radius);
  border-top-right-radius: var(--border-radius);
  font-family: var(--primary-font-family);
  font-size: 24px;
  font-weight: bold;
  display: flex;
  align-items: center;
`;

const Status = styled.div`
  flex: 0 0 50%;
  background-color: ${props =>
    `var(--color-${props.delays ? "delay-background" : "white"})`};
  color: ${props => `var(--color-${props.delays ? "delay-text" : "black"})`};
  border-bottom-left-radius: var(--border-radius);
  border-bottom-right-radius: var(--border-radius);
  font-size: 24px;

  padding: 16px;
  text-align: center;
  .slick-slider {
    height: 100%;
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    .slick-arrow {
      background-color: ${props => `var(--color-${props.id})`};
      padding: 8px;
      width: 32px;
      height: 32px;
      border-radius: 50%;

      display: flex !important;
      height: 48px;
      width: 48px;
      align-items: center;
      justify-content: center;
      z-index: 3;
      transition: opacity 0.3s ease-in-out;
      svg {
        height: 16px;
        width: 16px;
      }
      &:before {
        display: none;
      }
      &.slick-next {
        right: 0;
        opacity: ${({ currentSlide, slideLength }) =>
          currentSlide === slideLength - 1 ? 0.6 : 1};
      }
      &.slick-prev {
        left: 0;
        opacity: ${({ currentSlide }) => (currentSlide === 0 ? 0.6 : 1)};
      }
    }
    .slick-list {
      width: 100%;
    }
  }
`;

/* eslint-disable-next-line */
const NextArrow = ({ className, onClick }) => (
  /* eslint-disable-next-line */
  <div className={className} onClick={onClick}>
    <FontAwesomeIcon
      icon={["fad", "arrow-right"]}
      color="var(--color-white)"
      size="1x"
    />
  </div>
);

/* eslint-disable-next-line */
const PrevArrow = ({ className, onClick }) => (
  /* eslint-disable-next-line */
  <div className={className} onClick={onClick}>
    <FontAwesomeIcon
      icon={["fad", "arrow-left"]}
      color="var(--color-white)"
      size="1x"
    />
  </div>
);

const Card = ({ line }) => {
  const [currentSlide, setCurrentSlide] = useState(0);

  const settings = {
    infinite: false,
    speed: 500,
    slidesToShow: 1,
    slidesToScroll: 1,
    nextArrow: <NextArrow />,
    prevArrow: <PrevArrow />,
    afterChange: current => setCurrentSlide(current),
  };

  return (
    <Wrapper>
      <Name id={line.id}>{line.name}</Name>
      <Status
        delays={
          line.lineStatuses.filter(status => status.statusSeverity > 0).length >
          0
        }
        id={line.id}
        currentSlide={currentSlide}
        slideLength={line.lineStatuses.length}
      >
        {/* eslint-disable-next-line */}
        <Slider {...settings}>
          {line.lineStatuses.map(status => (
            <div key={status.id}>{status.statusSeverityDescription}</div>
          ))}
        </Slider>
      </Status>
    </Wrapper>
  );
};

Card.propTypes = {
  line: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    lineStatuses: PropTypes.array,
  }).isRequired,
};

export default Card;
