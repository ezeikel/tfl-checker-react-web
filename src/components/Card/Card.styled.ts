import styled from "styled-components";

type StatusProps = {
  id: string;
  delays: boolean;
  currentSlide: any;
  slideLength: number;
};

type NameProps = {
  id: string;
};

export const Wrapper = styled.section`
  display: flex;
  flex-direction: column;
  height: 300px;
  box-shadow: var(--box-shadow);
  border-radius: var(--border-radius);
`;

export const Name = styled.div<NameProps>`
  flex: 0 0 50%;
  background-color: ${({ id }) => `var(--color-${id})`};
  color: var(--color-white);
  display: flex;
  align-content: center;
  justify-content: center;
  border-top-left-radius: var(--border-radius);
  border-top-right-radius: var(--border-radius);
  font-family: var(--primary-font-family);
  font-size: 24px;
  font-weight: bold;
  display: flex;
  align-items: center;
`;

export const Status = styled.div<StatusProps>`
  flex: 0 0 50%;
  background-color: ${({ delays }) =>
    `var(--color-${delays ? "delay-background" : "white"})`};
  color: ${({ delays }) => `var(--color-${delays ? "delay-text" : "black"})`};
  border-bottom-left-radius: var(--border-radius);
  border-bottom-right-radius: var(--border-radius);
  font-size: 24px;
  padding: 16px;
  text-align: center;
  .slick-slider {
    height: 100%;
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    .slick-arrow {
      background-color: ${({ id }) => `var(--color-${id})`};
      padding: 8px;
      width: 32px;
      height: 32px;
      border-radius: 50%;
      display: flex !important;
      height: 48px;
      width: 48px;
      align-items: center;
      justify-content: center;
      z-index: 3;
      transition: opacity 0.3s ease-in-out;
      svg {
        height: 16px;
        width: 16px;
      }
      &:before {
        display: none;
      }
      &.slick-next {
        right: 0;
        opacity: ${({ currentSlide, slideLength }) =>
          currentSlide === slideLength - 1 ? 0.6 : 1};
      }
      &.slick-prev {
        left: 0;
        opacity: ${({ currentSlide }) => (currentSlide === 0 ? 0.6 : 1)};
      }
    }
    .slick-list {
      width: 100%;
    }
  }
`;
