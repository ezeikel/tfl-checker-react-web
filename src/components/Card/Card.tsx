import { useState } from "react";
import Slider from "react-slick";
import "slick-carousel/slick/slick.css";
import "slick-carousel/slick/slick-theme.css";
import { Wrapper, Name, Status } from "./Card.styled";
import NextArrow from "../NextArrow/NextArrow";
import PreviousArrow from "../PreviousArrow/PreviousArrow";

type CardProps = {
  line: {
    id: string;
    name: string;
  };
};

const Card = ({ line }: CardProps) => {
  const [currentSlide, setCurrentSlide] = useState(0);

  const settings = {
    infinite: false,
    speed: 500,
    slidesToShow: 1,
    slidesToScroll: 1,
    nextArrow: <NextArrow />,
    prevArrow: <PreviousArrow />,
    afterChange: (current) => setCurrentSlide(current),
  };

  return (
    <Wrapper>
      <Name id={line.id}>{line.name}</Name>
      <Status
        delays={
          line.lineStatuses.filter((status) => status.statusSeverity > 0)
            .length > 0
        }
        id={line.id}
        currentSlide={currentSlide}
        slideLength={line.lineStatuses.length}
      >
        {/* eslint-disable-next-line */}
        <Slider {...settings}>
          {line.lineStatuses.map((status) => (
            <div key={status.id}>{status.statusSeverityDescription}</div>
          ))}
        </Slider>
      </Status>
    </Wrapper>
  );
};

export default Card;
