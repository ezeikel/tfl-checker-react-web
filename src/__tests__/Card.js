import React from 'react';
import { render } from '@testing-library/react';
import Card from '../components/Card';

let mockProps;

beforeEach(() => {
  mockProps = {
    line: {
      id: "central",
      name: "Central",
      lineStatuses: [{
        statusSeverity: 3,
        statusSeverityDescription: "Part Suspended"
      }],
    }
  };
});

it('renders without crashing', () => {
  const { container } = render(
    <Card {...mockProps} /> // eslint-disable-line react/jsx-props-no-spreading
  );
  expect(container).toMatchSnapshot();
});

it('renders without crashing', () => {
  const { container } = render(
    <Card {...mockProps} /> // eslint-disable-line react/jsx-props-no-spreading
  );
  expect(container).toMatchSnapshot();
});

it('renders the correct line name', () => {
  const { getByText } = render(
    <Card {...mockProps} /> // eslint-disable-line react/jsx-props-no-spreading
  );
  expect(getByText('Central')).toBeInTheDocument();
});

it('renders the correct line status', () => {
  const { getByText } = render(
    <Card {...mockProps} /> // eslint-disable-line react/jsx-props-no-spreading
  );
  expect(getByText('Part Suspended')).toBeInTheDocument();
});