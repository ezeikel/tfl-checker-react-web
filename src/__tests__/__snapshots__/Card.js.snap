// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <section
    class="sc-AxjAm bfWoqw"
  >
    <div
      class="sc-AxirZ jqiuBa"
      id="central"
    >
      Central
    </div>
    <div
      class="sc-AxiKw kTKAoU"
      id="central"
    >
      <div
        class="slick-slider slick-initialized"
      >
        
        <div
          class="slick-list"
        >
          <div
            class="slick-track"
            style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
          >
            <div
              aria-hidden="false"
              class="slick-slide slick-active slick-current"
              data-index="0"
              style="outline: none; width: 0px;"
              tabindex="-1"
            >
              <div>
                <div
                  style="width: 100%; display: inline-block;"
                  tabindex="-1"
                >
                  Part Suspended
                </div>
              </div>
            </div>
          </div>
        </div>
        
        
      </div>
    </div>
  </section>
</div>
`;

exports[`renders without crashing 2`] = `
<div>
  <section
    class="sc-AxjAm bfWoqw"
  >
    <div
      class="sc-AxirZ jqiuBa"
      id="central"
    >
      Central
    </div>
    <div
      class="sc-AxiKw kTKAoU"
      id="central"
    >
      <div
        class="slick-slider slick-initialized"
      >
        
        <div
          class="slick-list"
        >
          <div
            class="slick-track"
            style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
          >
            <div
              aria-hidden="false"
              class="slick-slide slick-active slick-current"
              data-index="0"
              style="outline: none; width: 0px;"
              tabindex="-1"
            >
              <div>
                <div
                  style="width: 100%; display: inline-block;"
                  tabindex="-1"
                >
                  Part Suspended
                </div>
              </div>
            </div>
          </div>
        </div>
        
        
      </div>
    </div>
  </section>
</div>
`;
